### DISCUSSION TITLE
Release Schedule

### DISCUSSION THREAD

**Discussion by NiklasGustafsson** [xamarin/TorchSharp#479] [Announcements]
Adding a thread to discuss upcoming releases of TorchSharp.

**Comment by NiklasGustafsson** [comment]
**0.95.4**

I intend to make a release 0.95.4 to NuGet on December 3. 

This will contain the DisposeScope addition, as well as some bug fixes and a couple of additional LR schedulers. If we can get the DataLoader PR completed, it may go in, too.

It will be the last NuGet release or TorchSharp this year, as I am taking vacation from December 6 onward. I will attend to critical bugs only during that time.

**Comment by NiklasGustafsson** [comment]
__0.96.0__

I intend to make another release to NuGet today, for version 0.96.0.

It contains fixes to a few bugs, and some breaking changes:

1. 'Module.Eval' is now 'Module.eval' and 'Module.Train' is 'Module.train'

2. 'Module.parameters' and 'Module.named_parameters()' now return IEnumerable<T> instead of T[]. This is closer to the PyTorch API and potentially slightly more efficient, but it might break any API that assumes it's getting an array.

**Comment by NiklasGustafsson** [comment]
**0.96.1**

Another release will be coming up, after issues #510 and #495 are resolved. No date yet, but likely the week of February 14.

**Reply by NiklasGustafsson** [reply]
It's not likely to be happening this week. There's a huge PR coming up that needs to be reviewed, and that will take time.

**Reply by GeorgeS2019** [reply]
This is the [PR](https://github.com/dotnet/TorchSharp/pull/523)

**Reply by NiklasGustafsson** [reply]
Don Syme created a release with some critical F#-related fixes. The big bug release will be 0.96.3 (I messed up the 0.96.2 release).

**Comment by NiklasGustafsson** [comment]
**0.96.4**

The big feature in this release will be support for .NET FX 4.7.2 and later.

That said, this release will come once open issues 551-559 are addressed. Volunteers to help with some of these issues would be appreciated, but it will get done no matter what.

Pytorch 1.11 has been released, but we will deal with that in a later release of TorchSharp.

**Reply by NiklasGustafsson** [reply]
I just sent 0.96.4 to NuGet. Please see the [Release Notes](https://github.com/dotnet/TorchSharp/blob/main/RELEASENOTES.md) for details.

**Comment by NiklasGustafsson** [comment]
I just release 0.96.5 to NuGet. The previous release had some issues when used in VS Code that were not caught before 0.96.4 was released. This release fixes those.

**Comment by NiklasGustafsson** [comment]
0.96.6 was just released to NuGet. There are a small number of bug fixes, and the addition of a few "standard" models under TorchVision.

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.96.7

This release is now available on NuGet. Here's the summary of changes:

__Dependency Changes:__

This version integrates with the libtorch 1.10.0 backend. API updates to follow.<br/>

__API Changes:__

Strong name signing of the TorchSharp library to allow loading it in .NET Framework strongly name signed apps.<br/>
Added the 'META' device type, which can be used to examine the affect of shape from tensor operations without actually doing any computations.<br/>
Added a few methods from the torch.nn.utils namespace.<br/>
Add torch.stft() and torch.istft()

__Fixed Bugs:__

#567 pad missing the choice to fill at start or end<br/>

**Comment by NiklasGustafsson** [comment]
Just released to NuGet:

## NuGet Version 0.96.8

__Breaking Changes:__

This release contains a fix to inadvertent breaking changes in 0.96.7, related to Tensor.str(). This fix is itself breaking, in that it breaks any code that relies on the order of
arguments to str() introduced in 0.96.7. However, since the pre-0.96.7 argument order makes more sense, we're taking this hit now rather than keeping the inconvenient order in 0.96.7.

__Fixed Bugs:__

#618 TorchSharp.Modules.Normal.sample() Expected all tensors [...]<br/>
#621 torch.roll missing<br/>
#629 Missing dependency in 0.96.7 calling TorchSharp.torchvision.datasets.MNIST<br/>
#632 gaussian_nll_loss doesn't work on GPU<br/>

__API Changes:__

Add torchaudio.datasets.YESNO().
Added torch.from_array() API to create a tensor from an arbitry-dimension .NET array.<br/>
Added torch.tensor() overloads for most common dimensions of .NET arrays: ndim = [1,2,3,4]<br/>
Added the most significant API additions from Pytorch 1.11.<br/>
Added two torchaudio APIs.<br/>
Added 'decimals' argument to Tensor.round()<br/>
Changed tensor.str() to undo the breaking change in 0.96.7<br/>
Added torch.std_mean()<br/>

**Comment by NiklasGustafsson** [comment]
Just released on NuGet:

## NuGet Version 0.97.0

__Fixed Bugs:__

#653:Tensor.to(Tensor) doesn't change dtype of Tensor.<br/>

__API Changes__:

<b>Add ability to load and save TorchScript modules created using Pytorch</b><br/>
Add torch.utils.rnn<br/>
Add torchvision.io<br/>
Add Tensor.trace() and torch.trace() (unrelated to torch.jit.trace)<br/>
Add Tensor.var and Tensor.var_mean<br/>
Add torchaudio.datasets.SPEECHCOMMANDS<br/>
Add torchaudio.Resample()<br/>

**Comment by NiklasGustafsson** [comment]
Just released on NuGet:

## NuGet Version 0.97.1

This release is made shortly after 0.97.0, since it adresses a serious performance issue when creating large tensors from .NET arrays.

__Fixed Bugs:__

#670 Tensor allocation insanely slow for from_array()<br/>

__API Changes__:

RNN, LSTM, GRU support PackedSequence<br/>
Add element-wise comparison methods of torch class.<br/>
Fix clamp and (non)quantile method declarations<br/>
Implementing isnan()<br/>
Added torchaudio.models.Tacotron2()<br/>

**Comment by NiklasGustafsson** [comment]
Soon on NuGet:

## NuGet Version 0.97.2

__Breaking Changes:__

This release contains a breaking change related to `torch.tensor()` and `torch.from_array()` which were not adhering to the semantics of the Pytorch equivalents (`torch.from_numpy()` in the case of `torch.from_array()`).

With this change, there will be a number of different APIs to create a tensor form a .NET array. The most significant difference between them is whether the underlying storage is shared, or whether a copy is made. Depending on the size of the input array, copying can take orders of magnitude more time in creation than sharing storage, which is done in constant time (a few μs).

The resulting tensors may be reshaped, but not resized.

```C#
// Never copy:
public static Tensor from_array(Array input)

// Copy only if dtype or device arguments require it:
public static Tensor frombuffer(Array input, ScalarType dtype, long count = -1, long offset = 0, bool requiresGrad = false, Device? device = null)
public static Tensor as_tensor(Array input,  ScalarType? dtype = null, Device? device = null)
public static Tensor as_tensor(Tensor input, ScalarType? dtype = null, Device? device = null)

// Always copy:
public static Tensor as_tensor(IList<<VARIOUS TYPES>> input,  ScalarType? dtype = null, Device? device = null)
public static Tensor tensor(<<VARIOUS TYPES>> input, torch.Device? device = null, bool requiresGrad = false)
```

__Fixed Bugs:__

#670 Better align methods for creating tensors from .NET arrays with Pytorch APIs. _This is the breaking change mentioned earlier.
#679 The default value of onesided or torch.istft() is not aligned with PyTorch

__API Changes__:

Added torch.nn.init.trunc_normal_
Added index_add, index_copy, index_fill
Added torch.frombuffer()
Added torch.fft.hfft2, hfftn, ihfft2, ihfftn
Adding SequentialLR to the collection of LR schedulers.
Add 'training' flag to functional dropout methods.
Add missing functions to torchaudio.functional
Adding TestOfAttribute to unit tests

**Comment by NiklasGustafsson** [comment]
Just released to NuGet:

## NuGet Version 0.97.3

__Fixed Bugs:__

#694 torch.log10() computes torch.log()
#691 torch.autograd.backward()
#686 torch.nn.functional.Dropout() doesn't have the training argument.

__API Changes__:

Add `repeat_interleave()`
Add torch.broadcast_shapes()
Added meshgrid, mT, mH, and H
Added additional distributions.
Add dct and mu-law to torchaudio
Added torchvision -- sigmoid_focal_loss()
Update the arguments of `dropout()` in `Tacotron2`
Add static function for `all()`, `any()`, `tile()`, `repeat_interleave()`.
Add an implementation of the ReduceLROnPlateau learning rate scheduler.

**Comment by NiklasGustafsson** [comment]
I release 0.97.4 to NuGet, but it's a bad release -- the Google.Protobuf dependency is not captured, so you have to install it manually in every project that uses Tensorboard. I will release 0.97.5 to fix it.

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.97.5

__Fixed Bugs:__

#715 How to implement the following code

__API Changes__:

Add functional normalizations.
Added torch.utils.tensorboard.SummaryWriter. Support for scalars only.

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.97.6

__Breaking Changes:__

This release changes TorchSharp.torchvision from a namespace to a static class. This will break any using directives that assumes that it is a namespace.

__Fixed Bugs:__

#719 ResNet maxpool
#730 Sequential.Add
#729 Changing torchvision namespace into a static class?

__API Changes__:

Adding 'append()' to torch.nn.Sequential
Adding torch.numel() and torch.__version__
Adding modifiable global default for tensor string formatting

**Comment by NiklasGustafsson** [comment]
On its way to NuGet:

## NuGet Version 0.98.0

__Breaking Changes:__

Some parameter names were changed to align with PyTorch. This affects names like 'dimension,' 'probability,' and 'keepDims' and will break code that is passing these parameters by name.

Module.to(), cpu(), and cuda() were moved to a static class for extension methods. This means that it is necessary to have a 'using TorchSharp;' (C#) or 'open TorchSharp' (F#) in each file using them.

Doing so (rather than qualifying names with 'TorchSharp.') was already recommended as a best practice, since such a using/open directive will allows qualified names to align with the PyTorch module hierarchy.

__Loss functions are now aligned with the PyTorch APIs.__ This is a major change and the reason for incrementing the minor version number. The most direct consequence is that losses are modules rather than delegates, which means you need to call .forward() to actually compute the loss. Also, the factories are in torch.nn rather than torch.nn.functional and have the same Pascal-case names as the corresponding types. The members of the torch.nn.functional static class are now proper immediate loss functions, whereas the previous ones returned a loss delegate.

__Generic Module base class.__ The second major change is that Module is made type-safe with respect to the `forward()` function. Module is now an abstract base class, and interfaces `IModule<T,TResult>`, `IModule<T1,T2,TResult>`,... are introduced to define the signature of the `forward()` function. For most custom modules, this  means that the base class has to be changed to `Module<Tensor,Tensor>`, but some modules may require more significant changes.

ScriptModule follows this pattern, but this version introduces `ScriptModule<T...,TResult>` base classes, with corresponding `torch.jit.load<T...,TResult>()` static factory methods.

__Fixed Bugs:__

#323 forward() should take a variable-length list of arguments
#558 Fix deviation from the Pytorch loss function/module APIs
#742 Ease of use: Module.to method should be generic T -> T
#743 Ease of use: module factories should have dtype and device
#745 Executing a TorchScript that returns multiple values, throws an exception
#744 Some of functions with inconsistent argument names
#749 functional.linear is wrong
#761 Stateful optimizers should have support for save/load from disk.
#771 Support more types for ScriptModule

__API Changes__:

Module.to(), cpu(), and cuda() were redone as extension methods. The virtual methods to override, if necessary, are now named '_to'. A need to do so should be extremely rare.
Support for saving and restoring hyperparameters and state of optimizers
Loss functions are now Modules rather than delegates.
Custom modules should now use generic versions as base classes.
ScriptModule supports calling methods other than forward()
Added torch.jit.compile().

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.98.1

__Breaking Changes:__

`TorchVision` and `TorchAudio` have beeen moved into their own NuGet packages, which need to be added to any project using their APIs.

ModuleList and ModuleDict are now generic types, taking the module type as the type parameter. torch.nn.ModuleDict() will return a ModuleDict<Module>, which torch.nn.ModuleDict<T>() will return a ModuleDict<T>, where T must be a Module type.

__Fixed Bugs:__

#568 Overloads for Named Tensors
#765 Support invoking ScriptModule methods
#775 torch.jit.load: support specifying a target device
#792 Add SkiaSharp-based default imager for torchvision.io

__API Changes__:

Generic ModuleDict and ModuleList
Added torchaudio.transforms.GriffinLim
Added support for named tensors
Added default dim argument value for 'cat'

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.98.2

__Breaking Changes__:

The .NET 5.0 is no longer supported. Instead, .NET 6.0 is the minimum version. .NET FX 4.7.2 and higher are still supported.

__API Changes__:

Support 'null' as input and output to/from TorchScript.
Added support for label smoothing in CrossEntropyLoss.
Added torchaudio.transforms.MelSpectrogram().
Adding squeeze_()
Adding older-style tensor factories -- IntTensor, FloatTensor, etc.

__Fixed Bugs__:

#783 Download progress bar missing
#787 torch.where(condition) → tuple of LongTensor function missing
#799 TorchSharp.csproj refers Skia

__Source Code Cleanup__:

Moved P/Invoke declarations into dedicated class.
Added C# language version to all .csproj files.

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.98.3

This is an "emergency" release, to address some serious package reference issues introduced in 0.98.2

__Fixed Bugs__:

MultiStepLR scheduler was not computing the next LR correctly.
Fixed incorrect version in package reference.
Added missing package references to TorchVision manifest.

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.99.0

This is an upgrade to libtorch 1.13. It also moves the underlying CUDA support to 11.7 from 11.3, which means that all the libtorch-cuda-* packages have been renamed.

__Breaking Changes__:

See API Changes.<br/>

__API Changes__:

Removed Tensor.lstsq, paralleling PyTorch. Use torch.linalg.lstsq, instead. This is a breaking change.
Added 'left' Boolean argument to `torch.linalg.solve().

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.99.1

__Breaking Changes__:

The options to the ASGD, Rprop, and RMSprop optimizers have been changed to add a 'maximize' flag. This means that saved state dictionaries for these optimizers will not carry over.

The return type of Sequential.append() has changed from 'void' to 'Sequential.' This breaks binary compatibility, but not source compat.

__API Changes__:

Added a number of 1.13 APIs under `torch.special`
Added a `maximize` flag to the ASGD, Rprop and RMSprop optimizers.
Added PolynomialLR scheduler
The return type of Sequential.append() has changed from 'void' to 'Sequential.'
Added 1-dimensional array overloads for `torch.as_tensor()`

__Fixed Bugs__:

#836 Categorical seems to be miscalculated
#838 New Bernoulli get "Object reference not set to an instance of an object."
#845 registered buffers are being ignored in move model to device
#851 tensor.ToString(TorchSharp.TensorStringStyle.Numpy)
#852 The content returned by torch.nn.Sequential.append() is inconsistent with the official

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.99.2

__API Changes__:

Adding 'maximize' argument to the Adadelta optimizer
Adding linalg.ldl_factor and linalg.ldl_solve
Adding a couple of missing APIs (see #872)
Adding SoftplusTransform
Support indexing and slicing of Sequential
Adding ToNDArray() to TensorAccessor

__Fixed Bugs__:

#870 nn.AvgPool2d(kernel_size=3, stride=2, padding=1) torchsharp not support padding
#872 Tensor.masked_fill_(mask, value) missing
#877 duplicate module parameters called named_parameters() while load model by cuda
#888 THSTensor_meshgrid throws exception

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.99.3

__API Changes__:

Fixing misspelling of 'DetachFromDisposeScope,' deprecating the old spelling.
Adding allow_tf32
Adding overloads of Module.save() and Module.load() taking a 'Stream' argument.
Adding torch.softmax() and Tensor.softmax() as aliases for torch.special.softmax()
Adding torch.from_file()
Adding a number of missing pointwise Tensor operations.
Adding select_scatter, diagonal_scatter, and slice_scatter
Adding torch.set_printoptions
Adding torch.cartesian_prod, combinations, and cov.
Adding torch.cdist, diag_embed, rot90, triu_indices, tril_indices

__Fixed Bugs__:

#913 conv = nn.Conv2d(c1, 1, 1, bias=False).requires_grad_(False)
#910 nn.Module.modules is missing
#912 nn.Module save and state_ dict method error

**Comment by TheBlackPlague** [comment]
The last release was on Feb 22nd. It's Apr 2nd now. When's the next release? @NiklasGustafsson

**Comment by NiklasGustafsson** [comment]
I was hoping to get #936 merged before releasing 0.99.4, but I can do another release later this week sometime if it's important.

**Reply by TheBlackPlague** [reply]
I can wait.

**Reply by NiklasGustafsson** [reply]
Done!

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.99.4

__Breaking Changes__:

There were some changes to the binary format storing optimizer state. This means that any such state generated before updating to this version is invalid and will likely result in a runtime error.

__API Changes__:

Adding torch.tensordot
Adding torch.nn.Fold and Unfold modules.
Adding `Module.call()` to all the `Module<T...>` classes. This wraps `Module.forward()` and allows hooks to be registered. `Module.forward()` is still available, but the most general way to invoke a module's logic is through `call()`.
Adding tuple overloads for all the padding-related modules.
Adding support for exporting optimizer state from PyTorch and loading it in TorchSharp

__Fixed Bugs__:

#842 How to use `register_forward_hook`?
#940 Missing `torch.searchsorted`
#942 `nn.ReplicationPad1d(long[] padding)` missing
#943 `LRScheduler.get_last_lr` missing
#951 DataLoader constructor missing `drop_last` parameter
#953 TensorDataset is missing
#962 Seed passed to `torch.random.manual_seed(seed)` is unused
#949 Passing optimizer state dictionary from PyTorch to TorchSharp
#971 std results are inconsistent

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.99.5

__API Changes__:

Added Tensorboard support for histograms, images, video, and text.

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.99.6

There were a few problems with 0.99.5 that required a quick follow-up fix release. Specifically, the NuGet package listed a dependency on a different version of System.Memory than TorchSharp was actually dependent on, which caused problems. Also, a version of `torch.squeeze()` was defined with default arguments that caused ambiguity when no arguments were passed. Both were unacceptable.

__Breaking Changes__:

There was a second version of `torch.squeeze()` with incorrect default arguments. It has now been removed.

__API Changes__:

Removed incorrect `torch.squeeze()` method.
Adding two-tensor versions of `min()` and `max()`

__Fixed Bugs__:

#984 Conversion from System.Index to TensorIndex is missing
#987 Different versions of System.Memory between build and package creation.

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.100.0

Updated backend binaries to libtorch v2.0.1.

Updated the NuGet metadata to use a license expression rather than a reference to a license file. This will help with automated license checking by users.

Note that when running with CUDA on Windows, you will get this warning from deep within the PyTorch runtime, which does not seem to impact anything:

```
[W C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\torch\csrc\jit\codegen\cuda\interface.cpp:47] Warning: Loading nvfuser library failed with: error in LoadLibrary for nvfuser_codegen.dll. WinError 126: The specified module could not be found.
 (function LoadingNvfuserLibrary)
```

If you find what and where it has an impact on execution, please report it as an issue.

__Breaking Changes__:

With v2.0.1, `torch.istft()` expects complex numbers in the input tensor.

__API Changes__:

#989 Adding anomaly detection APIs to `torch.autograd`

__Fixed Bugs__:

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.100.1

__Breaking Changes__:

TorchSharp no longer supports any .NET Core versions prior to 6.0. .NET FX version support is still the same: 4.7.2 and up.

__API Changes__:

Added operator functionality to Torchvision, but roi are still missing.
Added support for additional types related to TorchScript modules. Scripts can now return lists of lists and tuples of lists and tuples, to an arbitrary level of nesting.
Scripts can now accept lists of Tensors.

__Bug Fixes__:

#1001 Issue with resnet50, resnet101, and resnet152

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.100.2

__API Changes__:

Add torchvision.datasets.CelebA()
Add support for properly formatting Tensors in Polyglot notebooks without the 'Register' call that was necessary before.

__Bug Fixes__:

#1014 AdamW.State.to() ignores returns.
#999 Error in Torchsharp model inference in version 0.100.0

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.100.3

This release is primarily, but not exclusively, focused on fixing bugs in distributions and adding a few new ones.

__Breaking Changes__:

The two main arguments to `torch.linalg.solve()` and `torch.linalg.solve_ex()` were renamed 'A' and 'B' to align with PyTorch.

__API Changes__:

Adding torch.linalg.solve_triangular()
Adding torch.distributions.MultivariateNormal
Adding torch.distributions.NegativeBinomial
Adding in-place versions of `Tensor.triu()` and `Tensor.tril()`
Adding torch.linalg.logsigmoid() and torch.nn.LogSigmoid
A number of distributions were missing the `mode` property.
Adding a C#-like string formatting style for tensors.

__Bug Fixes__:

TorchVision `rotate()`, `solarize()` and `invert()` were incorrectly implemented.
Fixed bug in Bernoulli's `entropy()` and `log_prob()` implementations.
Fixed bug in Cauchy's `log_prob()` implementation.
Fixed several bugs in HalfCauchy and HalfNormal.
The Numpy-style string formatting of tensors was missing commas between elements

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.100.4

__Breaking Changes__:

The constructor for TensorAccessor is now `internal`, which means that the only way to create one is to use the `data<T>()` method on Tensor. This was always the intent.

__API Changes__:

Tensor.randperm_out() deprecated. 
torch.randperm accepts 'out' argument 
Adding PReLU module. 
Adding scaled_dot_product_attention. 
The constructor for TensorAccessor was made `internal` 
torchvision.utils.save_image implemented 
torchvision.utils.make_grid implemented 
torchvision.transforms.RandAugment implemented 

__Bug Fixes__:

Fixed torch.cuda.synchronize() method 
Suppress runtime warning by setting align_corners to 'false' 
Fixed argument validation bug in Grayscale 
#1056: Access violation with TensorAccessor.ToArray - incompatible data types 
#1057: Memory leak with requires_grad

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.100.5

__Breaking Changes__:

Inplace operators no longer create an alias, but instead return 'this'. This change will impact any code that explicitly calls `Dispose` on a tensor after the operation.

__Bug Fixes__:

#1041 Running example code got error in Windows 10
#1064 Inplace operators create an alias
#1084 Module.zero_grad() does not work
#1089 max_pool2d overload creates tensor with incorrect shape

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.100.6

__Bug Fixes__:

ScriptModule: adding `forward` and the ability to hook.
Update to SkiaSharp 2.88.6 to avoid the libwebp vulnerability.
#1105: Dataset files get written to the wrong directory.
#1116: Gradient null for simple calculation.

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.100.7

Even though the last release was only a week ago, I concluded that there was a critical mass of important bug fixes, especially making it work better with F# and fixing the iterable-style DataLoader issue.

__Breaking Changes__:

DataLoader should no longer be created using `new` -- instead, the overall pattern is followed, placing the classes in `TorchSharp.Modules` and the factories in the static class. This will break any code that creates a DataLoader, but can be fixed by:

1. Removing the `new` in `new torch.utils.data.DataLoader(...)`
2. Adding a `using TorchSharp.Modules` (C#) or `open TorchSharp.Modules` (F#) to files where `DataLoader` is used as a type name.

__API Changes__:

Adding an `IterableDataset` abstract class, and making `TensorDataset` derive from it.
Moving the `DataLoader` class to `TorchSharp.Modules` and adding DataLoader factories.
#1092: got error when using DataLoader
#1069: Implementation of torch.sparse_coo_tensor for sparse tensor creation
Renamed `torch.nn.functional.SiLU` -> `torch.nn.functional.silu`
Added a set of generic `Sequential` classes.

__Bug Fixes__:

#1083: Compiler rejects scalar operand due to ambiguous implicit conversion

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.101.0

This is an upgrade to libtorch 2.1.0. It also moves the underlying CUDA support to 12.1 from 11.7, which means that all the libtorch-cuda-* packages have been renamed. Please update your CUDA driver to one that support CUDA 12.1 before moving to this version of TorchSharp.

__API Changes__:

Enhanced `Module.load` function to return matching status of parameters in non-strict mode via an output dictionary.
Introduced attribute-based parameter naming for module state dictionaries, allowing custom names to override default field names.

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.101.1

This is a fast-follower bug fix release, addressing persistent issues with stability of using TorchScript from TorchSharp. Given the significance of this fix, a same-day release is made.

__Bug Fixes__:

#1047 Torchscript execution failures (hangs, access violation, Fatal error. Internal CLR fatal error. (0x80131506) )

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.101.2

__API Changes__:

Added extension method `ScalarType.ElementSize()` to get the size of each element of a given ScalarType.
Added methods for loading and saving individual tensors with more overloads.
Added 'persistent' flag to register_buffer()

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.101.3

__Breaking Changes__:

The base `OptimizerState` class was modified and includes two changes:

1. Custom optimizer state objects derived from `OptimizerState` must now explicitly pass the related `torch.nn.Parameter` object to the `OptimizerState` base constructor to maintain correct linkage.
2. Custom state objects must implement an Initialize function. This function is responsible for initializing the properties of the state. Note that this function can be called as a re-intialization, so proper disposal of the previous tensor objects should be handled.

__API Changes__:

Introduced `InferenceMode`, a block-based scoping class for optimizing TorchSharp model inference by disabling gradient computation and enhancing performance. 
Added `Tensor.to_type()` conversion aliases for short, half, bfloat16, cfloat, and cdouble. 
Added `Module.to()` conversion aliases for all the scalar types. 
All distribution classes now implement IDisposable. 

__Bug Fixes__:

#1154 : `mu_product` was not initialized in `NAdam` optimizer 
#1170 : Calling `torch.nn.rnn.utils.pad_packed_sequence` with a CUDA tensor and unsorted_indices threw an error 
#1172 : `optim.LoadStateDict` from an existing `StateDictionary` updated to make sure to copy value and to the right device. 
#1176 : When specific `Optimizers` load in a conditional tensor, made sure to copy to the right device. 
#1174 : Loading CUDA tensor from stream threw an error 
#1179 : Calling `Module.to()` with the `ParameterList` and `ParameterDict` module didn't move the parameters stored in the field. 
#1148 : Calling `Module.to()` shouldn't be differentiable 
#1126 : Calling `ScriptModule.to()` doesn't move attributes 
#1180 : Module.to(ScalarType) has restrictions in PyTorch which aren't restricted in TorchSharp.

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.101.4

A fast-follow release addressing a regression in v0.101.3

__Bug Fixes__:

#1185 : Incomplete transfer of module to device (only with 0.101.3)

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.101.5

Another release just before 2024. There will be no more releases in 2023. It addresses and important bug in `Module.to()`

__Bug Fixes__:

#1191 : Having trouble moving a module from one GPU to another with gradients.

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.101.6

__API Changes__:

#1223: Missing prod function torch.prod or a.prod() where a is Tensor 
#1201: How to access the attributes of a model? 
#1094: ScriptModule from Stream / ByteArray 
#1149: Implementation for `torch.autograd.functional.jacobian` to compute Jacobian of a function 
Implemenation for a custom `torch.autograd.Function` class 

__Bug Fixes__:

#1198: CUDA not available when calling backwards before using CUDA 
#1200: Bugs in torch.nn.AvgPool2d and torch.nn.AvgPool3d methods.

**Comment by NiklasGustafsson** [comment]
## NuGet Version 0.102.0

This release upgrades the libtorch backend to v2.2.0.

__Breaking Changes__:

The Ubuntu builds are now done on a 22.04 version of the OS. This may (or may not) affect TorchSharp use on earlier versions.
The default value for the `end_factor` argument in the constructor for `LinearLR` was changed to 1.0 to match PyTorch.
Any code that checks whether a device is 'CUDA' and does something rather than checking that it isn't 'CPU' will now fail to work, since there is now support for the 'MPS' device on MacOS.

__API Changes__:

#652: Apple Silicon support .
#1219: Added support for loading and saving tensors that are >2GB.

__Bug Fixes__:

Fixed LinearLR scheduler calculation with misplaced parentheses.
Added `get_closed_form_lr` to scheduler to match PyTorch behavior when specifying epoch in `.step()`

**Comment by NiklasGustafsson** [comment]
# NuGet Version 0.102.1

__Breaking Changes__:

The `kernelSize` parameter in the function and class of `AvgPool1D` was renamed to `kernel_size` to match PyTorch naming.
The `stride` parameter in the `torch.nn.functional.avg_pool1d` call now defaults to `kernelSize` instead of 1, to match the PyTorch behavior.


__Bug Fixes__:

`module.load_state_dict()` throws error for in-place operation on a leaf variable that requires grad.
#1250 cstr and npstr for 0d tensors
#1249 torch.nn.functional.avg_pool1d is not working correctly
`module.load()` with streams which don't read the requested # of bytes throws error.
 #1246 Issue running in notebook on Apple Silicon

**Comment by NiklasGustafsson** [comment]
# NuGet Version 0.102.2

This was a fast-follower release based on a pretty important bug fix that caused memory leaks.

__Bug Fixes__:

#1257 InverseMelScale in NewDisposeScope doesn't dispose tensors<br/>

**Comment by NiklasGustafsson** [comment]
# NuGet Version 0.102.3

__Breaking Changes__:

The 'paddingMode' parameter of convolution has been changed to 'padding_mode', and the 'outputPadding' is now 'output_padding'.

__API Changes__:

#1243 `fuse_conv_bn_weights` and `fuse_linear_bn_weights` are added.
#1274 ConvTranspose3d does not accept non-uniform kernelSize/stride values.

**Comment by NiklasGustafsson** [comment]
# NuGet Version 0.102.4

__Breaking Changes__:

Correct `torch.finfo`. (`torch.set_default_dtype`, `Categorical.entropy`, `_CorrCholesky.check`, `Distribution.ClampProbs`, `FisherSnedecor.rsample`, `Gamma.rsample`, `Geometric.rsample`, `distributions.Gumbel`, `Laplace.rsample`, `SigmoidTransform._call` and `SigmoidTransform._inverse` are influenced.)

__API Changes__:

#1284 make `torch.unique` and `torch.unique_consecutive` public.

**Comment by NiklasGustafsson** [comment]
# NuGet Version 0.102.5

__Breaking Changes__:

`torchvision.dataset.MNIST` will try more mirrors. The thrown exception might be changed when it fails to download `MNIST`, `FashionMNIST` or `KMNIST`.
`ObjectDisposedException` will now be thrown when trying to use a disposed dispose scopes.
The constructor of dispose scopes is no longer `public`. Use `torch.NewDisposeScope` instead.

__API Changes__:

#1317 How to set default device type in torchsharp.
#1314 Grant read-only access to DataLoader attributes
#1313 Add 'non_blocking' argument to tensor and module 'to()' signatures.
#1291 `Tensor.grad()` and `Tensor.set_grad()` have been replaced by a new property `Tensor.grad`.
A potential memory leak caused by `set_grad` has been resolved.
`Include` method of dispose scopes has been removed. Use `Attach` instead.
Two more `Attach` methods that accepts `IEnumerable<IDisposable>`s and arrays as the parameter have been added into dispose scopes.
A new property `torch.CurrentDisposeScope` has been added to provide the ability to get the current dispose scope.
Add module hooks that take no input/output arguments, just the module itself.

__Bug Fixes__:

#1300 `Adadelta`, `Adam` and `AdamW` will no longer throw `NullReferenceException` when `maximize` is `true` and `grad` is `null`.
torch.normal` will now correctly return a leaf tensor.
New options `disposeBatch` and `disposeDataset` have been added into `DataLoader`.
The default collate functions will now always dispose the intermediate tensors, rather than wait for the next iteration.

__Bug Fixes__:

`TensorDataset` will now keep the aliases detached from dispose scopes, to avoid the unexpected disposal.
`DataLoaderEnumerator` has been completely rewritten to resolve the unexpected shuffler disposal, the ignorance of `drop_last`, the incorrect count of worker, and the potential leak cause by multithreading.
#1303 Allow dispose scopes to be disposed out of LIFO order.

**Comment by NiklasGustafsson** [comment]
# NuGet Version 0.103.0

Move to libtorch 2.4.0.
Jettison MacOS / Intel support (libtorch no longer supports it). 
Accumulated bug fixes.

__Breaking Changes__:

When creating a tensor from a 1-D array, and passing in a shape, there is now an ambiguity between the IList and Memory overloads of `torch.tensor()`. The ambiguity is resolved by removing the `dimensions` argument if it is redundant, or by an explicit cast to IList if it is not.

__Bug Fixes__:

#1359 torch.nn.functional.l1_loss computes a criterion with the MSE, not the MAE.
#1334 MultivariateNormal.log_prob() exception in TorchSharp but works in pytorch.

__API Changes__:

#1326 Allow arrays used to create tensors to be larger than the tensor. Create tensors from a Memory instance.<br/>
